docs/design/assembly-loading.md

```diff
@@ -0,0 +1,98 @@
+
+XML descriptors from the static reference closure are processed whether or not the embedding assembly is marked.
+
+### Dynamic assembly references not discovered
+
+Later steps do not load additional assemblies, with rare exceptions. Dynamic references (from reflection, or from special attributes) are only looked up in already-loaded assemblies, so the linker is unable to find assemblies only referenced dynamically, producing incorrect outputs which are missing the code required by such references. This problem is partially mitigated for `DynamicDependencyAttribute` by a step that pre-scans for assemblies referenced from instances of this attribute, but it does not recursively handle `DynamicDependencyAttribute` instances in these referenced assemblies.
+
+## .NET6 behavior
+
+### Assembly references loaded lazily
+
+Assemblies will be resolved and loaded lazily, whether by name or from a metadata reference. We will avoid loading referenced assemblies up-front, instead relying on various APIs that all ultimately call `IAssemblyResolver.Resolve` to load assemblies as needed. Trying to `Resolve` an assembly name (using `IAssemblyResolver` or a `LinkContext` helper) or a Cecil member reference (`MethodReference`, `ExportedType`, etc...) will continue to have predictable assembly loading behavior - it will either find a cached result, or look for and possibly load a new assembly. Unused references will typically not be loaded at all, removing the need to process potentially large amounts of unused code. Just resolving an assembly will not trigger any additional processing or recursive assembly resolution.
+
+### Additional processing done lazily
+
+Additional processing will be done lazily (not triggered by `Resolve`), running on a single assembly at a time.
+
+- Type info base/override methods will be built per assembly as requested
+
+  The overrides tracked for a method may be incomplete, and will be updated as new assemblies are loaded. This is compatible with the current design of `MarkStep` which re-processes virtual methods, but this could be optimized in the future.
+
+- XML from an assembly will be processed only if the assembly is marked
+
+  Embedded XML descriptors will logically be considered dependencies of the assembly, not global dependencies.
+
+- Constant propagation and branch elimination will run per marked assembly
+
+  This leaves room to optimize constant propagation to run at a more granular level inline with the marking logic.
+
+### Breaking changes
+
+We will restrict the embedded XML so that it may only modify the containing assembly. This is to prevent cases where a (possibly lazily loaded) assembly modifies code in another assembly that we have already processed, breaking assumptions and creating logical inconsistencies. XML passed on the command-line will continue to work as they do today.
```

Comment:
This is correct, but I would still point out that descriptors will be allowed to bring in/mark parts of other assemblies, including referencing new assemblies.
This also adds a restriction on descriptors in that they can only add code, they cannot remove or modify code (now and in the future).